utils/compression/zstd_compressor.go

```diff
@@ -0,0 +1,58 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package compression
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"math"
+
+	"github.com/DataDog/zstd"
+)
+
+var _ Compressor = (*zstdCompressor)(nil)
+
+func NewZstdCompressor(maxSize int64) (Compressor, error) {
+	if maxSize == math.MaxInt64 {
+		// "Decompress" creates "io.LimitReader" with max size + 1:
+		// if the max size + 1 overflows, "io.LimitReader" reads nothing
+		// returning 0 byte for the decompress call
+		// require max size < math.MaxInt64 to prevent int64 overflows
+		return nil, ErrInvalidMaxSizeCompressor
+	}
+
+	return &zstdCompressor{
+		maxSize: maxSize,
+	}, nil
+}
+
+type zstdCompressor struct {
+	maxSize int64
+}
+
+func (z *zstdCompressor) Compress(msg []byte) ([]byte, error) {
+	if int64(len(msg)) > z.maxSize {
+		return nil, fmt.Errorf("%w: (%d) > (%d)", ErrMsgTooLarge, len(msg), z.maxSize)
+	}
+	return zstd.Compress(nil, msg)
+}
+
+func (z *zstdCompressor) Decompress(msg []byte) ([]byte, error) {
+	reader := zstd.NewReader(bytes.NewReader(msg))
+	defer reader.Close()
+
+	// We allow [io.LimitReader] to read up to [z.maxSize + 1] bytes, so that if
+	// the decompressed payload is greater than the maximum size, this function
+	// will return the appropriate error instead of an incomplete byte slice.
+	limitReader := io.LimitReader(reader, z.maxSize+1)
+	decompressed, err := io.ReadAll(limitReader)
+	if err != nil {
+		return nil, err
+	}
+	if int64(len(decompressed)) > z.maxSize {
+		return nil, fmt.Errorf("%w: (%d) > (%d)", ErrDecompressedMsgTooLarge, len(decompressed), z.maxSize)
```

Comment:
Why do we return an error here? At this point we've already decompressed the payload so it seems like a waste to drop this message - i think it makes sense to limit the size of the thing we're compressing but it feels strange to limit the size of the result of the decompression

Reply:
if we don't do this error handling then i think we can also not do the weird extra byte allocation in the earlier line

Reply:
We did not decompress the message in this case. We stopped decompressing the message because it may have been a zip bomb.